src/Analysis/Engine/Impl/Interpreter/Ast/AstLazyMember.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.PythonTools.Interpreter.Ast {
+    class AstLazyMember : ILazyMember {
+        private readonly Func<IMember> _getter;
+        private IMember _value;
+
+        [DebuggerStepThrough]
+        public AstLazyMember(Func<IMember> getter) {
+            _getter = getter;
+        }
+        public PythonMemberType MemberType => PythonMemberType.Unknown;
+
+        public string Name => "lazy";
+
+        [DebuggerStepThrough]
+        public IMember Get() => _value ?? (_value = _getter());
```

Comment:
Should we check if `_getter` is called inside `AnalysisQueue`?

Reply:
To make sure it is resolved?

Reply:
And to avoid unexpected concurrent calls. `DDG` is executed only inside `AnalysisQueue`. Do we expect `ILazyMember.Get()` to be called not from analyzer?

Reply:
Yes, it can be called at the end of AST walks and also is called in tests without analyzer present